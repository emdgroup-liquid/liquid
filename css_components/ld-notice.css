.ld-notice,:host{--ld-notice-error-icon-col:var(--ld-thm-error);--ld-notice-info-icon-col:var(--ld-thm-primary);--ld-notice-warning-icon-col:var(--ld-thm-warning);background:var(--ld-col-wht);border-radius:var(--ld-br-l);box-shadow:var(--ld-shadow-stacked);color:var(--ld-col-neutral-800);display:block;font:var(--ld-typo-body-m);padding:var(--ld-sp-24);padding-left:calc(var(--ld-sp-24) + var(--ld-sp-40));position:relative;}.ld-notice ::slotted(a),.ld-notice a,:host ::slotted(a),:host a{color:var(--ld-thm-primary);text-decoration:underline}.ld-notice ::slotted(code),.ld-notice code,:host ::slotted(code),:host code{background-color:var(--ld-col-neutral-010);border-radius:var(--ld-br-s);box-shadow:inset 0 0 0 var(--ld-sp-1) var(--ld-col-neutral-100);padding:0 var(--ld-sp-4)}.ld-notice--error .ld-notice__headline,.ld-notice--error .ld-notice__icon,:host(.ld-notice--error) .ld-notice__headline,:host(.ld-notice--error) .ld-notice__icon{color:var(--ld-notice-error-icon-col)}.ld-notice--error .ld-notice__icon,:host(.ld-notice--error) .ld-notice__icon{transform:rotate(180deg)}.ld-notice--info .ld-notice__headline,.ld-notice--info .ld-notice__icon,:host(.ld-notice--info) .ld-notice__headline,:host(.ld-notice--info) .ld-notice__icon{color:var(--ld-notice-info-icon-col)}.ld-notice--warning,:host(.ld-notice--warning){background:var(--ld-thm-warning-highlight);}.ld-notice--warning .ld-notice__icon,:host(.ld-notice--warning) .ld-notice__icon{--ld-icon-secondary-col:var(--ld-col-neutral-900);color:var(--ld-notice-warning-icon-col)}.ld-notice--warning .ld-notice__headline,:host(.ld-notice--warning) .ld-notice__headline{color:var(--ld-col-neutral-900)}.ld-notice__headline{margin:0 0 var(--ld-sp-12)}.ld-notice__icon{left:var(--ld-sp-24);position:absolute;top:var(--ld-sp-24)}