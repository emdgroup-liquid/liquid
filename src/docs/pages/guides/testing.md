---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

