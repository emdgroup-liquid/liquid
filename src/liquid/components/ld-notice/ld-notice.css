:host,
.ld-notice {
  --ld-notice-error-icon-col: var(--ld-thm-error);
  --ld-notice-info-icon-col: var(--ld-thm-primary);
  --ld-notice-warning-icon-col: var(--ld-thm-warning);
  --ld-notice-success-icon-col: var(--ld-thm-success);

  background: var(--ld-layer-2);
  border-radius: var(--ld-br-l);
  box-shadow: var(--ld-shadow-stacked);
  color: var(--ld-col-neutral-800);
  display: block;
  font: var(--ld-typo-body-m);
  padding: var(--ld-sp-24);
  padding-left: calc(var(--ld-sp-24) + var(--ld-sp-40));
  position: relative;

  ::slotted(a),
  a {
    color: var(--ld-thm-primary);
    text-decoration: underline;
  }

  ::slotted(code),
  code {
    background-color: var(--ld-col-neutral-010);
    border-radius: var(--ld-br-s);
    box-shadow: inset 0 0 0 var(--ld-sp-1) var(--ld-col-neutral-100);
    padding: 0 var(--ld-sp-4);
  }
}

:host(.ld-notice--error),
.ld-notice--error {
  ::slotted(ld-icon),
  ::slotted(.ld-icon),
  .ld-notice__icon,
  .ld-notice__headline {
    color: var(--ld-notice-error-icon-col);
  }

  .ld-notice__icon {
    transform: rotate(180deg);
  }
}

:host(.ld-notice--info),
.ld-notice--info {
  ::slotted(ld-icon),
  ::slotted(.ld-icon),
  .ld-notice__icon,
  .ld-notice__headline {
    color: var(--ld-notice-info-icon-col);
  }
}

:host(.ld-notice--warning),
.ld-notice--warning {
  background: var(--ld-thm-warning-highlight);

  ::slotted(ld-icon),
  ::slotted(.ld-icon),
  .ld-notice__icon {
    --ld-icon-secondary-col: var(--ld-col-neutral-900);
    color: var(--ld-notice-warning-icon-col);
  }

  .ld-notice__headline {
    color: var(--ld-col-neutral-900);
  }
}

:host(.ld-notice--success),
.ld-notice--success {
  ::slotted(ld-icon),
  ::slotted(.ld-icon),
  .ld-notice__icon {
    color: var(--ld-notice-success-icon-col);
  }

  .ld-notice__headline {
    color: var(--ld-notice-success-icon-col);
  }
}

.ld-notice__headline {
  margin: 0 0 var(--ld-sp-12);
}

::slotted(ld-icon),
::slotted(.ld-icon),
.ld-notice__icon {
  left: var(--ld-sp-24);
  position: absolute;
  top: var(--ld-sp-24);
}
