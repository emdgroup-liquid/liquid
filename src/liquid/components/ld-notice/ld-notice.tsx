import { Component, h, Prop, Host, Element } from '@stencil/core'
import { HTMLStencilElement } from '@stencil/core/internal'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part headline - `ld-typo` element used for the headline
 * @part icon - Image tag used for the icon
 */
@Component({
  assetsDirs: ['assets'],
  tag: 'ld-notice',
  styleUrl: 'ld-notice.css',
  shadow: true,
})
export class LdNotice {
  @Element() hostElement: HTMLStencilElement

  /** Headline of the notice. */
  @Prop() headline?: string

  /** Mode of the notice. */
  @Prop() mode: 'error' | 'info' | 'warning' | 'success' = 'info'

  render() {
    return (
      <Host class={`ld-notice ld-notice--${this.mode}`}>
        <slot name="custom-icon">
          <ld-icon
            class="ld-notice__icon"
            name={
              this.mode === 'success'
                ? 'ld-input-message-success'
                : 'ld-input-message-info'
            }
            part="icon"
            size="lg"
          />
        </slot>
        {this.headline && (
          <ld-typo
            class="ld-notice__headline"
            variant="h4"
            tag="p"
            part="headline"
          >
            {this.headline}
          </ld-typo>
        )}
        <slot></slot>
      </Host>
    )
  }
}
