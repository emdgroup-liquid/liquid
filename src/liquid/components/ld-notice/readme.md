---
eleventyNavigation:
  key: Notice
  parent: Components
layout: layout.njk
title: Notice
permalink: components/ld-notice/
---

# ld-notice

<link rel="stylesheet" href="css_components/ld-notice.css">
<link rel="stylesheet" href="css_components/ld-icon.css">
<link rel="stylesheet" href="css_components/ld-typo.css">

This component is meant to be used in conjunction with the [`ld-input`](components/ld-input/) and the [`ld-label`](components/ld-label/) component. Please reffer to the [`ld-input`](components/ld-input/) docs for further usage examples.

## Examples

### As error message

{% example %}
<ld-notice headline="An error occurred" mode="error">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Metus pellentesque facilisi nunc iaculis. Laoreet eget eu lacus cursus odio quam ut elementum. Faucibus cursus in placerat enim non senectus. In molestie volutpat at sem bibendum ac id. Suspendisse erat malesuada vulputate et congue blandit in erat ornare. Rhoncus interdum.</ld-notice>

<!-- CSS component -->

<div class="ld-notice ld-notice--error">
  <!-- Note that you can use an img element with the class ld-notice__icon here, as well. -->
  <svg class="ld-notice__icon ld-icon ld-icon--lg" width="14" height="14" viewBox="0 0 14 14" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path fill-rule="evenodd" clip-rule="evenodd" d="M7 14C10.866 14 14 10.866 14 7C14 3.13401 10.866 0 7 0C3.13401 0 0 3.13401 0 7C0 10.866 3.13401 14 7 14Z" fill="currentColor"/>
    <path fill-rule="evenodd" clip-rule="evenodd" d="M7.18234 11.0254C6.79228 11.0254 6.48657 10.9147 6.26518 10.6933C6.0438 10.472 5.93311 10.1662 5.93311 9.77618V6.12335C5.93311 5.99685 6.0069 5.93359 6.15449 5.93359H6.89771C7.28776 5.93359 7.59348 6.04428 7.81487 6.26567C8.03625 6.48705 8.14694 6.79277 8.14694 7.18283V10.8357C8.14694 10.9622 8.07315 11.0254 7.92556 11.0254H7.18234Z" fill="var(--ld-icon-secondary-col)"/>
    <ellipse cx="6.99977" cy="3.80007" rx="1.06667" ry="1.06667" fill="var(--ld-icon-secondary-col)"/>
  </svg>
  <p class="ld-notice__headline ld-typo--h4">An error occurred</p>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Metus pellentesque facilisi nunc iaculis. Laoreet eget eu lacus cursus odio quam ut elementum. Faucibus cursus in placerat enim non senectus. In molestie volutpat at sem bibendum ac id. Suspendisse erat malesuada vulputate et congue blandit in erat ornare. Rhoncus interdum.
</div>
{% endexample %}

### As info message

{% example %}
<ld-notice headline="Information">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Metus pellentesque facilisi nunc iaculis. Laoreet eget eu lacus cursus odio quam ut elementum. Faucibus cursus in placerat enim non senectus. In molestie volutpat at sem bibendum ac id. Suspendisse erat malesuada vulputate et congue blandit in erat ornare. Rhoncus interdum.</ld-notice>

<!-- CSS component -->

<div class="ld-notice ld-notice--info">
  <!-- Note that you can use an img element with the class ld-notice__icon here, as well. -->
  <svg class="ld-notice__icon ld-icon ld-icon--lg" width="14" height="14" viewBox="0 0 14 14" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path fill-rule="evenodd" clip-rule="evenodd" d="M7 14C10.866 14 14 10.866 14 7C14 3.13401 10.866 0 7 0C3.13401 0 0 3.13401 0 7C0 10.866 3.13401 14 7 14Z" fill="currentColor"/>
    <path fill-rule="evenodd" clip-rule="evenodd" d="M7.18234 11.0254C6.79228 11.0254 6.48657 10.9147 6.26518 10.6933C6.0438 10.472 5.93311 10.1662 5.93311 9.77618V6.12335C5.93311 5.99685 6.0069 5.93359 6.15449 5.93359H6.89771C7.28776 5.93359 7.59348 6.04428 7.81487 6.26567C8.03625 6.48705 8.14694 6.79277 8.14694 7.18283V10.8357C8.14694 10.9622 8.07315 11.0254 7.92556 11.0254H7.18234Z" fill="var(--ld-icon-secondary-col)"/>
    <ellipse cx="6.99977" cy="3.80007" rx="1.06667" ry="1.06667" fill="var(--ld-icon-secondary-col)"/>
  </svg>
  <p class="ld-notice__headline ld-typo--h4">Information</p>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Metus pellentesque facilisi nunc iaculis. Laoreet eget eu lacus cursus odio quam ut elementum. Faucibus cursus in placerat enim non senectus. In molestie volutpat at sem bibendum ac id. Suspendisse erat malesuada vulputate et congue blandit in erat ornare. Rhoncus interdum.
</div>
{% endexample %}

### As warning message

{% example %}
<ld-notice headline="Warning!" mode="warning">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Metus pellentesque facilisi nunc iaculis. Laoreet eget eu lacus cursus odio quam ut elementum. Faucibus cursus in placerat enim non senectus. In molestie volutpat at sem bibendum ac id. Suspendisse erat malesuada vulputate et congue blandit in erat ornare. Rhoncus interdum.</ld-notice>

<!-- CSS component -->

<div class="ld-notice ld-notice--warning">
  <!-- Note that you can use an img element with the class ld-notice__icon here, as well. -->
  <svg class="ld-notice__icon ld-icon ld-icon--lg" width="14" height="14" viewBox="0 0 14 14" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path fill-rule="evenodd" clip-rule="evenodd" d="M7 14C10.866 14 14 10.866 14 7C14 3.13401 10.866 0 7 0C3.13401 0 0 3.13401 0 7C0 10.866 3.13401 14 7 14Z" fill="currentColor"/>
    <path fill-rule="evenodd" clip-rule="evenodd" d="M7.18234 11.0254C6.79228 11.0254 6.48657 10.9147 6.26518 10.6933C6.0438 10.472 5.93311 10.1662 5.93311 9.77618V6.12335C5.93311 5.99685 6.0069 5.93359 6.15449 5.93359H6.89771C7.28776 5.93359 7.59348 6.04428 7.81487 6.26567C8.03625 6.48705 8.14694 6.79277 8.14694 7.18283V10.8357C8.14694 10.9622 8.07315 11.0254 7.92556 11.0254H7.18234Z" fill="var(--ld-icon-secondary-col)"/>
    <ellipse cx="6.99977" cy="3.80007" rx="1.06667" ry="1.06667" fill="var(--ld-icon-secondary-col)"/>
  </svg>
  <p class="ld-notice__headline ld-typo--h4">Warning!</p>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Metus pellentesque facilisi nunc iaculis. Laoreet eget eu lacus cursus odio quam ut elementum. Faucibus cursus in placerat enim non senectus. In molestie volutpat at sem bibendum ac id. Suspendisse erat malesuada vulputate et congue blandit in erat ornare. Rhoncus interdum.
</div>
{% endexample %}


<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description                                              | Type                             | Default     |
| ---------- | ---------- | -------------------------------------------------------- | -------------------------------- | ----------- |
| `headline` | `headline` | Headline of the notice.                                  | `string`                         | `undefined` |
| `key`      | `key`      | for tracking the node's identity when working with lists | `string \| number`               | `undefined` |
| `mode`     | `mode`     | Mode of the notice.                                      | `"error" \| "info" \| "warning"` | `'info'`    |
| `ref`      | `ref`      | reference to component                                   | `any`                            | `undefined` |


## Shadow Parts

| Part         | Description                             |
| ------------ | --------------------------------------- |
| `"headline"` | `ld-typo` element used for the headline |
| `"icon"`     | Image tag used for the icon             |


## Dependencies

### Depends on

- [ld-icon](../ld-icon)
- [ld-typo](../ld-typo)

### Graph
```mermaid
graph TD;
  ld-notice --> ld-icon
  ld-notice --> ld-typo
  style ld-notice fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

 
