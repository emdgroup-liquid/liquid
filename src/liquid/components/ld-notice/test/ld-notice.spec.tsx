import { h } from '@stencil/core'
import { newSpecPage } from '@stencil/core/testing'
import { LdNotice } from '../ld-notice'

describe('ld-notice', () => {
  it('renders default', async () => {
    const page = await newSpecPage({
      components: [LdNotice],
      template: () => <ld-notice>Lorem ipsum dolor sit amet.</ld-notice>,
    })
    expect(page.root).toMatchSnapshot()
  })
  it('renders with headline', async () => {
    const page = await newSpecPage({
      components: [LdNotice],
      template: () => (
        <ld-notice headline="Headline">Lorem ipsum dolor sit amet.</ld-notice>
      ),
    })
    expect(page.root).toMatchSnapshot()
  })
  it('renders with mode', async () => {
    const page = await newSpecPage({
      components: [LdNotice],
      template: () => (
        <ld-notice mode="error">Lorem ipsum dolor sit amet.</ld-notice>
      ),
    })
    expect(page.root).toMatchSnapshot()
  })
})
