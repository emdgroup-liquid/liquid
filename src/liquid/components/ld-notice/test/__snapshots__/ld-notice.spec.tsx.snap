// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-notice renders default 1`] = `
<ld-notice class="ld-notice ld-notice--info">
  <mock:shadow-root>
    <slot name="custom-icon">
      <ld-icon class="ld-notice__icon" name="ld-input-message-info" part="icon" size="lg"></ld-icon>
    </slot>
    <slot></slot>
  </mock:shadow-root>
  Lorem ipsum dolor sit amet.
</ld-notice>
`;

exports[`ld-notice renders with custom icon 1`] = `
<ld-notice class="ld-notice ld-notice--success">
  <mock:shadow-root>
    <slot name="custom-icon">
      <ld-icon class="ld-notice__icon" name="ld-input-message-success" part="icon" size="lg"></ld-icon>
    </slot>
    <slot></slot>
  </mock:shadow-root>
  <ld-icon name="placeholder" size="lg" slot="custom-icon"></ld-icon>
  Lorem ipsum dolor sit amet.
</ld-notice>
`;

exports[`ld-notice renders with headline 1`] = `
<ld-notice class="ld-notice ld-notice--info">
  <mock:shadow-root>
    <slot name="custom-icon">
      <ld-icon class="ld-notice__icon" name="ld-input-message-info" part="icon" size="lg"></ld-icon>
    </slot>
    <ld-typo class="ld-notice__headline" part="headline" tag="p" variant="h4">
      Headline
    </ld-typo>
    <slot></slot>
  </mock:shadow-root>
  Lorem ipsum dolor sit amet.
</ld-notice>
`;

exports[`ld-notice renders with mode 1`] = `
<ld-notice class="ld-notice ld-notice--error">
  <mock:shadow-root>
    <slot name="custom-icon">
      <ld-icon class="ld-notice__icon" name="ld-input-message-info" part="icon" size="lg"></ld-icon>
    </slot>
    <slot></slot>
  </mock:shadow-root>
  Lorem ipsum dolor sit amet.
</ld-notice>
`;
